pgrx/src/callconv.rs

```diff
@@ -9,96 +9,280 @@
+            Ret::Many(_, _) => unreachable!(),
+        }
+    }
+
+    unsafe fn prepare_call(_fcinfo: pg_sys::FunctionCallInfo) -> CallCx {
+        CallCx::WrappedFn(unsafe { pg_sys::CurrentMemoryContext })
+    }
+
+    unsafe fn into_context(self, _fcinfo: pg_sys::FunctionCallInfo) {
+        unimplemented!()
+    }
+    unsafe fn ret_from_context(_fcinfo: pg_sys::FunctionCallInfo) -> Ret<Self> {
+        unimplemented!()
+    }
+    unsafe fn finish_call(_fcinfo: pg_sys::FunctionCallInfo) {}
+}
+
+pub enum CallCx {
+    RestoreCx,
+    WrappedFn(pg_sys::MemoryContext),
+}
+
+fn prepare_value_per_call_srf(fcinfo: pg_sys::FunctionCallInfo) -> CallCx {
+    unsafe {
+        if srf_is_first_call(fcinfo) {
+            let fn_call_cx = pg_sys::init_MultiFuncCall(fcinfo);
+            CallCx::WrappedFn((*fn_call_cx).multi_call_memory_ctx)
+        } else {
+            CallCx::RestoreCx
+        }
+    }
+}
+
+unsafe impl<'a, T> ReturnShipping for SetOfIterator<'a, T>
+where
+    T: ReturnShipping,
+{
+    type Item = <Self as Iterator>::Item;
+    unsafe fn prepare_call(fcinfo: pg_sys::FunctionCallInfo) -> CallCx {
+        prepare_value_per_call_srf(fcinfo)
+    }
+
+    fn label_ret(self) -> Ret<Self> {
+        let mut iter = self;
+        let ret = iter.next();
+        match ret {
+            None => Ret::Zero,
+            Some(value) => Ret::Many(iter, value),
+        }
+    }
+
+    unsafe fn box_return(fcinfo: pg_sys::FunctionCallInfo, ret: Ret<Self>) -> pg_sys::Datum {
+        let value = match ret {
+            Ret::Zero => return empty_srf(fcinfo),
+            Ret::Once(value) => value,
+            Ret::Many(iter, value) => {
+                iter.into_context(fcinfo);
+                value
             }
+        };
+
+        unsafe {
+            let fcx = deref_fcx(fcinfo);
+            srf_return_next(fcinfo, fcx);
+            T::box_return(fcinfo, value.label_ret())
         }
+    }
 
+    unsafe fn into_context(self, fcinfo: pg_sys::FunctionCallInfo) {
         let fcx = deref_fcx(fcinfo);
-        // SAFETY: fcx.user_fctx was set earlier, immediately before or in a prior call
-        let setof_iterator = &mut *(*fcx).user_fctx.cast::<SetOfIterator<T>>();
+        unsafe {
+            let ptr = srf_memcx(fcx).leak_and_drop_on_delete(self);
+            // it's the first call so we need to finish setting up fcx
+            (*fcx).user_fctx = ptr.cast();
+        }
+    }
 
-        match setof_iterator.next() {
-            Some(datum) => {
-                srf_return_next(fcinfo, fcx);
-                datum.into_datum().unwrap_or_else(|| pg_return_null(fcinfo))
-            }
-            None => empty_srf(fcinfo),
+    unsafe fn ret_from_context(fcinfo: pg_sys::FunctionCallInfo) -> Ret<Self> {
+        let fcx = deref_fcx(fcinfo);
+        // SAFETY: fcx.user_fctx was set earlier, immediately before or in a prior call
+        let iter = &mut *(*fcx).user_fctx.cast::<TableIterator<T>>();
```

Comment:
uh that is extremely not correct
```suggestion
        let iter = &mut *(*fcx).user_fctx.cast::<SetOfIterator<'_, T>>();
```